### PR TITLE
Configurable default numeric type

### PR DESCRIPTION
**Pull request by bortolin**
Hi Davide,

The Parser correctly consider numbers as integers or doubles if they have decimal places just like the C# compiler does.

In some cases it may be useful to be able to configure the default type of numbers if no particular suffix is ​​specified: for example in financial calculations, where usually numbers are interpreted as decimal type.

Another user @MrGorki have write about this on this issue [https://github.com/davideicardi/DynamicExpresso/issues/35](https://github.com/davideicardi/DynamicExpresso/issues/35).

I have implemented DefaultNumberType settings for configure the default type of number (Int, Long, Float, Double, Decimal).

I have also update documentation about this new features to specify how to use.

I have not touch the default behavior of parser if the settings is not set or set to Default.

I hope this new feature can be useful to others who like me use this beautiful library in financial application.

Marco

### CODE REVIEW

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 1]
```diff
@@ -2333,7 +2356,17 @@ private void NextToken()
...
+						//RealLiteral if DefaultNumberType settings is set to real type
+						if (_defaultNumberType == DefaultNumberType.Single || _defaultNumberType == DefaultNumberType.Double || _defaultNumberType == DefaultNumberType.Decimal)
+                        {
+							t = TokenId.RealLiteral;
+                        }
+                        else
+                        {
+							//IntegerLiteral by default
+							t = TokenId.IntegerLiteral;
+						}
```
Comment (davideicardi) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:2368]: Minor code alignement issues

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 2]
```diff
@@ -703,14 +710,30 @@ private Expression ParseRealLiteral()
...
+				else if (_defaultNumberType == DefaultNumberType.Single)
+				{
+					if (float.TryParse(text, ParseLiteralDecimalNumberStyle, ParseCulture, out float f))
+						value = f;
+				}
+                else
+                {
+					if (double.TryParse(text, ParseLiteralDoubleNumberStyle, ParseCulture, out double d))
+						value = d;
+				}
```
Comment (davideicardi) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:735]: Minor code alignement issues

**Comment by davideicardi** [comment]
Thank you @bortolin ! I like the idea!

I will review the code in depth in the following days.

@metoule Given that you have recently touch that part of the code, can I ask your help on review it when you have some free time? Do you see problems?

File: test/DynamicExpresso.UnitTest/LiteralsTest.cs [hunk 3]
```diff
@@ -75,6 +75,217 @@ public void Numeric_Literals()
...
+			// m suffix (decimal)
+			Assert.AreEqual(5M, target.Eval("5M"));
+			Assert.AreEqual(254m, target.Eval("254m"));
+			Assert.AreEqual(45.232M, target.Eval("45.232M"));
+			Assert.AreEqual(45.232m, target.Eval("45.232m"));
+			Assert.AreEqual(.022M, target.Eval(".022M"));
+			Assert.AreEqual(.022m, target.Eval(".022m"));
+			Assert.AreEqual(-.022m, target.Eval("-.022m"));
+			Assert.AreEqual(-.022M, target.Eval("-.022M"));
+		}
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/LiteralsTest.cs:128]: Could you also add tests with the `u` suffix? For example:

`Assert.AreEqual(45L, target.Eval("45u"));` 

and also tests with the minus unary operator, such as:

`Assert.AreEqual(-45L, target.Eval("-45u"));`

**Comment by metoule** [comment]
I'd just add a few more tests, but otherwise it looks good to me :)

File: test/DynamicExpresso.UnitTest/LiteralsTest.cs (hunk 3 above)
Comment (bortolin) [reply test/DynamicExpresso.UnitTest/LiteralsTest.cs:132]: I have fixed

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 2 above)
Comment (bortolin) [reply src/DynamicExpresso.Core/Parsing/Parser.cs]: I have fixed

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 1 above)
Comment (bortolin) [reply src/DynamicExpresso.Core/Parsing/Parser.cs]: I have fixed

File: test/DynamicExpresso.UnitTest/LiteralsTest.cs [hunk 4]
```diff
@@ -75,6 +77,219 @@ public void Numeric_Literals()
 			Assert.AreEqual(-.022M, target.Eval("-.022M"));
 		}
```
Comment (davideicardi) [inline test/DynamicExpresso.UnitTest/LiteralsTest.cs:79]: I think it should be useful to add the following test:

```
		[Test]
		public void Numeric_Literals_DefaultTypes()
		{
			var target = new Interpreter();
			Assert.IsInstanceOf(typeof(System.Int32), target.Eval("81"));
			Assert.IsInstanceOf(typeof(System.Double), target.Eval("81.5"));
			Assert.IsInstanceOf(typeof(System.Int64), target.Eval("23423423423434"));
		}
```

To verify that by default numbers are interpreted as pre documentation.

**Review by davideicardi — CHANGES_REQUESTED** [review]
Thanks! I ask you two final changes. See my comments.

File: test/DynamicExpresso.UnitTest/LiteralsTest.cs (hunk 4 above)
Comment (bortolin) [reply test/DynamicExpresso.UnitTest/LiteralsTest.cs:79]: I have added the Numeric_Literals_DefaultTypes test.

File: src/DynamicExpresso.Core/Interpreter.cs [hunk 5]
```diff
@@ -102,6 +102,16 @@ public AssignmentOperators AssignmentOperators
 		#endregion
 
 		#region Options
+
+		/// <summary>
+		/// Allow to set de default numeric type when no suffix is specified (Int by default, Double if real number)
+		/// </summary>
+		/// <param name="defaultNumberType"></param>
+		public void SetDefaultNumberType(DefaultNumberType defaultNumberType)
```
[the remark this answers is not in the document]
Comment (bortolin) [reply src/DynamicExpresso.Core/Interpreter.cs]: Have you right, I fixed it.

**Review by davideicardi — APPROVED** [review]
Great! Thank you!

**Comment by davideicardi** [comment]
Published in version 2.6.0

**Pull request merged**